Compare jwt.x5t with signingKey.x5t when using validationParameters.IssuerSigningKeys in JwtSecurityTokenHandler.ResolveIssuerSigningKey

RojaEnnam opened the pull request:
Fixes #1339 

1. Adding functionality to compare X5t of each signingKey in IssuerSigningKeys to JWT token's `X5t.`
2. Adding a new instance of TheoryData in both JwtSecurityTokenHandlerTests and JsonWebTokenHandlerTests to include X509SigningCredential while creating a token, in order for jwtheader to assign token X5t value.

All the test runs are successful.

RojaEnnam committed:
JwtSecurityTokenHandler.ResolveIssuerSigningKey compare jwt.x5t with signingKey.x5t when using validationParameters.IssuerSigningKeys

brentschmaltz reviewed (commented):
This is the right idea. 

We have two handlers that work with JsonTokens.
1. JwtSecurityTokenHandler
2. JsonWebTokenHandler

The code in the JsonWebTokenHandler is fairly correct, see ResolveIssuerSigningKey. We may want to think about moving that code to a central spot, JwtTokenUtilities comes to mind, and create an internal method with a method signature perhaps like: (string kid, string x5t, SecurityToken jwt, TokenValidationParameters validationParameters).
The SecurityToken parameter is for extensibility that is sometimes helpful.

Keep in mind that JwtSecurityTokenHandler.ResolveIssuerSigningKey is virtual so it must be called, the utility method should be called from inside this method. Make sure same errors and exceptions are thrown.

Make sure to add tests, +ve and -ve.
See: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/How-We-Test

RojaEnnam committed:
This reverts commit 706dcc5e76f4e72250832c40cb269a0f85487f16.

Compare token and issuer signing keys x5t fields for resolving signing key during tokenvalidation

brentschmaltz commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:
```diff
@@ -1852,6 +1852,20 @@ public static TheoryData<JwtTheoryData> ValidateJwsTheoryData
                 {
+                     new JwtTheoryData
+                    {
+                        TestId = nameof(Default.AsymmetricJws) + "RequireSignedTokens",
```
nit. can you put a "_" between two names in the TestId (in all these test cases).
I know that was how the code was, but it will make reading the results better.

RojaEnnam replied:
done.

brentschmaltz commented on test/Microsoft.IdentityModel.TestUtils/Default.cs:
```diff
@@ -96,6 +96,10 @@ public static SigningCredentials AsymmetricSigningCredentials
         {
             get => new SigningCredentials(KeyingMaterial.DefaultX509SigningCreds_2048_RsaSha2_Sha2.Key, KeyingMaterial.DefaultX509SigningCreds_2048_RsaSha2_Sha2.Algorithm, KeyingMaterial.DefaultX509SigningCreds_2048_RsaSha2_Sha2.Digest);
         }
+        public static X509SigningCredentials X509AsymmetricSigningCredentials
```
nit: can you add a new line after the closing brace?

RojaEnnam replied:
done.

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -972,51 +972,7 @@ internal virtual SecurityKey ResolveIssuerSigningKey(JsonWebToken jwtToken, Toke
-            }
-
-            return null;
+            return JwtTokenUtilities.ResolveTokenSigningKey(jwtToken.Kid, jwtToken.X5t, jwtToken, validationParameters);
```
Nicely done :-)

RojaEnnam replied:
Thanks :)

brentschmaltz commented:
@RojaEnnam i had to rebase on top of dev the push -f.
